packages/api/src/webhooks/cannon.ts

```diff
@@ -0,0 +1,204 @@
+    }
+    return newInterval;
+  }
+
+  retry(event) {
+    event.lastInterval = this.calcBackoff(event.lastInterval);
+    event.status = "pending";
+    event.retries = event.retries ? event.retries + 1 : 1;
+    this.db.queue.updateMsg(event);
+  }
+
+  async _fireHook(
+    event: Queue,
+    webhook: Webhook,
+    sanitized: Stream,
+    user: User
+  ) {
+    console.log(`trying webhook ${webhook.name}: ${webhook.url}`);
+    let ips, urlObj, isLocal;
+    try {
+      urlObj = parseUrl(webhook.url);
+      if (urlObj.host) {
+        ips = await resolver.resolve4(urlObj.hostname);
+      }
+    } catch (e) {
+      console.error("error: ", e);
+      throw e;
+    }
+
+    // This is mainly useful for local testing
+    if (user.admin) {
+      isLocal = false;
+    } else {
+      try {
+        if (ips && ips.length) {
+          isLocal = isLocalIP(ips[0]);
+        } else {
+          isLocal = true;
+        }
+      } catch (e) {
+        console.error("isLocal Error", isLocal, e);
+        throw e;
+      }
+    }
+    if (isLocal) {
+      // don't fire this webhook.
+      console.log(
+        `webhook ${webhook.id} resolved to a localIP, url: ${webhook.url}, resolved IP: ${ips}`
+      );
+    } else {
+      console.log("preparing to fire webhook ", webhook.url);
+      // go ahead
+      let params = {
+        method: "POST",
+        headers: {
+          "content-type": "application/json",
+          "user-agent": "livepeer.com",
+        },
+        timeout: WEBHOOK_TIMEOUT,
+        body: JSON.stringify({
+          id: webhook.id,
+          event: webhook.event,
+          stream: sanitized,
+        }),
+      };
+
+      try {
+        logger.info(`webhook ${webhook.id} firing`);
+        let resp = await fetchWithTimeoutAndSleep(
+          webhook.url,
+          params,
+          event.lastInterval
+        );
+        if (resp.status >= 200 && resp.status < 300) {
+          // 2xx requests are cool.
+          // all is good
+          logger.info(`webhook ${webhook.id} fired successfully`);
+          await this.storeResponse(webhook, event, resp);
+          return true;
+        }
+        console.error(
+          `webhook ${webhook.id} didn't get 200 back! response status: ${resp.status}`
+        );
+        await this.storeResponse(webhook, event, resp);
+        // retry
+        this.retry(event);
+        return;
+        // return !webhook.blocking;
+      } catch (e) {
+        console.log("firing error", e);
+        // return !webhook.blocking;
+        return;
+      }
+    }
+  }
+
+  async storeResponse(
+    webhook: Webhook,
+    event: Queue,
+    resp: Response,
+    duration = 0
+  ) {
+    // TODO
+    // store response for each time a webhook fires.
+    await this.db.webhookResponse.create({
+      webhookId: webhook.id,
+      eventId: event.id,
+      statusCode: resp.status,
+      response: resp,
+    });
+  }
+
+  async onTrigger(event: Queue) {
+    if (!event) {
```

Comment:
We also need to check that event structure is valid (has all required fields at least).